File: randology-dump-master/reimpl/src/main/java/uk/ac/stir/randology/generator/LCG.java
package uk.ac.stir.randology.generator;

/**
 * An implementation of a Linear Congruential Generator with an arbitrary parameter set.
 * @author zak
 */
public class LCG implements RNG {
	
	private String name;
	
	private long multiplier;
	private long increment;
	private long seed;
	private long modulus;
	
	public LCG(long multiplier, long increment, long modulus, long seed) {
		this("LCG-" + multiplier + "-" + increment + "-" + modulus,multiplier,increment,modulus,seed);
	}
	
	public LCG(String name, long multiplier, long increment, long modulus, long seed) {
		this.multiplier = multiplier;
		this.increment = increment;
		this.modulus = modulus;
		this.seed = seed;
		setName(name);
	}
	
	public String getName() {
		return name;
	}
	
	private void setName(String name) {
		this.name = name;
	}
	
	public long nextLong() {
		seed = (seed*multiplier + increment) % modulus;
		return seed;
	}
	
    public long next64bits() {
    	long high = nextLong();
    	long med  = nextLong();
    	long low  = nextLong();
    	long result = (high << 33) ^ (med << 2) ^ (low << 62 >> 62);
    	return result;
    }
	
	public double nextDouble() {
		return Math.abs( (double)nextLong() / modulus );
	}

}

File: python-implementation/src/generator/LCG.py
from src.generator.RNG import RNG
import numpy as np


class LCG(RNG):

    def __init__(self, name: str, multiplier: np.int64, increment: np.int64, modulus: np.int64, seed: np.int64) -> None:
        RNG.__init__(self, name)
        self.multiplier = np.int64(multiplier)
        self.increment = np.int64(increment)
        self.modulus = np.int64(modulus)
        self.seed = np.int64(seed)

    def next_int(self) -> np.int64:
        self.seed = (self.seed * self.multiplier + self.increment) % self.modulus
        return self.seed

    def next_float(self) -> np.float64:
        return np.float64(abs(self.next_int() / self.modulus))

    def next_64_bits(self) -> np.int64:
        high = self.next_int()
        med = self.next_int()
        low = self.next_int()
        return np.int64((high << 33) ^ (med << 2) ^ (low << 62 >> 62))
